gtsfm/utils/geometry_comparisons.py

```diff
@@ -40,6 +43,122 @@ def align_rotations(aRi_list: List[Rot3], bRi_list: List[Rot3]) -> List[Rot3]:
     return [aRb.compose(bRi) for bRi in bRi_list]
 
 
+def ransac_align_poses_sim3_ignore_missing(
+    aTi_list_ref: List[Optional[Pose3]],
+    bTi_list_est: List[Optional[Pose3]],
+    num_iters: int = 1000,
+    delete_frac: float = DEFAULT_RANSAC_ALIGNMENT_DELETE_FRAC,
+    verbose: bool = False,
+) -> Tuple[List[Optional[Pose3]], Similarity3]:
+    """Align pose graphs by estimating a Similarity(3) transformation, while accounting for outliers in the pose graph.
+
+    Args:
+        aTi_list_ref: pose graph 1 (reference/target to align to).
+        bTi_list_est: pose graph 2 (to be aligned).
+        num_iters: number of RANSAC iterations to execture.
+        delete_frac: what percent of data to remove when fitting a single hypothesis.
+        verbose: whether to print out information about each iteration.
+
+    Returns:
+        best_aligned_bTi_list_est_full: transformed input poses previously "bTi_list" but now which
+            have the same origin and scale as reference (now living in "a" frame). Rather than
+            representing an aligned subset, this represents the entire aligned pose graph.
+        best_aSb: Similarity(3) object that aligns the two pose graphs (best among all hypotheses).
+    """
+    best_aSb = None
+    best_trans_error = float("inf")
+    best_rot_error = float("inf")
+
+    valid_idxs = [i for i, bTi in enumerate(bTi_list_est) if bTi is not None]
+    num_to_delete = math.ceil(delete_frac * len(valid_idxs))
+    if len(valid_idxs) - num_to_delete < 2:
+        # run without RANSAC! want at least 2 frame pairs for good alignment.
+        aligned_bTi_list_est, aSb = align_poses_sim3_ignore_missing(
+            aTi_list_ref, bTi_list_est
+        )
+        return aligned_bTi_list_est, aSb
+
+    # randomly delete some elements
+    for _ in range(num_iters):
+
+        aTi_list_ref_subset = copy.deepcopy(aTi_list_ref)
+        bTi_list_est_subset = copy.deepcopy(bTi_list_est)
+
+        # randomly delete 33% of the poses
+        delete_idxs = np.random.choice(a=valid_idxs, size=num_to_delete, replace=False)
+        for del_idx in delete_idxs:
+            bTi_list_est_subset[del_idx] = None
+
+        aligned_bTi_list_est, aSb = align_poses_sim3_ignore_missing(
+            aTi_list_ref_subset, bTi_list_est_subset
+        )
+
+        # evaluate inliers.
+        rot_error, trans_error, _, _ = compute_pose_errors_3d(aTi_list_ref, aligned_bTi_list_est)
+        if verbose:
+            print("Deleted ", delete_idxs, f" -> trans_error {trans_error:.1f}, rot_error {rot_error:.1f}")
+
+        if trans_error <= best_trans_error and rot_error <= best_rot_error:
```

Comment:
im not sure that is the best option. Basically if we have a hypothesis with a slightly worse rotation error but _much better_ translation error, we don't use it?

Reply:
Thanks for the review Akshay. Indeed there are other ways to implement this. In my experience thus far, this option works reasonably well.

Reply:
Are there any metrics or plots that might be more convincing? I am surprised, because it really depends on the order of hypotheses. And there is a clear reason why it would not work.